File: ta4j-core/src/main/java/org/ta4j/core/indicators/candles/ThreeBlackCrowsIndicator.java
/*
 * SPDX-License-Identifier: MIT
 */
package org.ta4j.core.indicators.candles;

import org.ta4j.core.Bar;
import org.ta4j.core.BarSeries;
import org.ta4j.core.Indicator;
import org.ta4j.core.indicators.CachedIndicator;
import org.ta4j.core.num.Num;

/**
 * Three black crows indicator.
 *
 * <p>
 * Matches the documented four-bar crow sequence at {@code index}: a bullish
 * white candle at {@code index - 3} followed by three black crows at
 * {@code index - 2}, {@code index - 1}, and {@code index}:
 *
 * <pre>
 * white candle at index - 3 is bullish
 * each crow at index k in {index - 2, index - 1, index} is bearish, has a
 *     very short lower shadow:
 *         lowerShadow(k) &lt;= 0.1 * priorAverageRange(k)
 * the first crow opens strictly below the white candle's high:
 *     open(index - 2) &lt; high(index - 3)
 * the second and third crows open within the previous body and close strictly
 *     below the previous close:
 *         open(k) &lt; open(k - 1) &amp;&amp; open(k) &gt; close(k - 1)
 *         close(k) &lt; close(k - 1)
 * </pre>
 *
 * <p>
 * The lower-shadow condition is inclusive, so a shadow exactly at the adaptive
 * threshold matches; the body-relative opens and declining closes are strict.
 * The short-shadow threshold is evaluated against the shared adaptive baseline
 * of the preceding {@code averagePeriod} candles.
 *
 * <p>
 * The default constructor uses a five-candle baseline ({@code averagePeriod} =
 * 5).
 *
 * <p>
 * The three crows are thresholded at offsets 0 through 2, so this indicator is
 * stable after {@code averagePeriod + 2} bars; the white candle at offset 3 is
 * geometry-only and needs no warm-up, only retained history. Before the
 * stability boundary, or when the retained history does not reach back to the
 * white candle, it returns {@code false}.
 *
 * <p>
 * Traditionally interpreted as a bearish reversal signal after an uptrend. This
 * indicator does not evaluate the preceding trend or any direction context; the
 * reversal context must be composed explicitly by the caller.
 *
 * @see <a href="http://www.investopedia.com/terms/t/three_black_crows.asp">
 *      http://www.investopedia.com/terms/t/three_black_crows.asp</a>
 */
public class ThreeBlackCrowsIndicator extends CandlePatternIndicator {

    /** Number of preceding candles averaged into the lower-shadow baseline. */
    private final int averagePeriod;

    private final transient Indicator<Num> lowerShadow;

    /**
     * Constructor with the default average period.
     *
     * @param series the bar series
     * @since 0.24.2
     */
    public ThreeBlackCrowsIndicator(BarSeries series) {
        super(CandleThresholdSupport.validateSeriesAndAveragePeriod(series,
                CandleThresholdSupport.DEFAULT_AVERAGE_PERIOD),
                CandleThresholdSupport.forSeries(series, CandleThresholdSupport.DEFAULT_AVERAGE_PERIOD));
        this.averagePeriod = CandleThresholdSupport.DEFAULT_AVERAGE_PERIOD;
        this.lowerShadow = thresholds.lowerShadow();
    }

    @Override
    int latestBaselineIndex(final int index) {
        return index - 2;
    }

    /**
     * Constructor with a custom average period.
     *
     * @param series        the bar series
     * @param averagePeriod the number of preceding candles averaged into the
     *                      lower-shadow baseline
     * @throws IllegalArgumentException if {@code averagePeriod} is below 1
     * @since 0.24.2
     */
    public ThreeBlackCrowsIndicator(BarSeries series, int averagePeriod) {
        super(CandleThresholdSupport.validateSeriesAndAveragePeriod(series, averagePeriod),
                CandleThresholdSupport.forSeries(series, averagePeriod));
        this.averagePeriod = averagePeriod;
        this.lowerShadow = thresholds.lowerShadow();
    }

    @Override
    protected Boolean calculate(int index) {
        BarSeries series = getBarSeries();
        if (index < getCountOfUnstableBars() || index - 3 < series.getBeginIndex()) {
            // We need 4 candles: 1 white, 3 black
            return false;
        }
        if (!Num.isFinite(lowerShadow.getValue(index - 2)) || !Num.isFinite(lowerShadow.getValue(index - 1))
                || !Num.isFinite(lowerShadow.getValue(index))) {
            return false;
        }

        Bar leadingBar = series.getBar(index - 3);
        if (!Num.isFinite(leadingBar.getOpenPrice()) || !Num.isFinite(leadingBar.getClosePrice())
                || !Num.isFinite(leadingBar.getHighPrice())) {
            return false;
        }

        return leadingBar.isBullish() && isBlackCrow(index - 2) && isBlackCrow(index - 1) && isBlackCrow(index);
    }

    @Override
    public int getCountOfUnstableBars() {
        return averagePeriod + 2;
    }

    /**
     * @param index the current bar/candle index
     * @return true if the current bar/candle is a black crow, false otherwise
     */
    private boolean isBlackCrow(int index) {
        Bar prevBar = getBarSeries().getBar(index - 1);
        Bar currBar = getBarSeries().getBar(index);
        if (currBar.isBearish()) {
            if (prevBar.isBullish()) {
                // First crow case
                return thresholds.isShortShadow(index, lowerShadow)
                        && currBar.getOpenPrice().isLessThan(prevBar.getHighPrice());
            } else {
                return thresholds.isShortShadow(index, lowerShadow) && isDeclining(index);
            }
        }
        return false;
    }

    /**
     * @param index the current bar/candle index
     * @return true if the current bar/candle is declining, false otherwise
     */
    private boolean isDeclining(int index) {
        Bar prevBar = getBarSeries().getBar(index - 1);
        Bar currBar = getBarSeries().getBar(index);
        final Num prevOpenPrice = prevBar.getOpenPrice();
        final Num prevClosePrice = prevBar.getClosePrice();
        final Num currOpenPrice = currBar.getOpenPrice();
        final Num currClosePrice = currBar.getClosePrice();

        // Opens within the body of the previous candle
        return currOpenPrice.isLessThan(prevOpenPrice) && currOpenPrice.isGreaterThan(prevClosePrice)
        // Closes below the previous close price
                && currClosePrice.isLessThan(prevClosePrice);
    }
}

File: ta4j-core/src/main/kotlin/org/ta4j/core/indicators/candles/ThreeBlackCrowsIndicator.kt
/**
 * The MIT License (MIT)
 *
 * Copyright (c) 2017-2022 Ta4j Organization & respective
 * authors (see AUTHORS)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
 * the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
package org.ta4j.core.indicators.candles

import org.ta4j.core.BarSeries
import org.ta4j.core.indicators.CachedIndicator
import org.ta4j.core.indicators.SMAIndicator
import org.ta4j.core.num.*

/**
 * Three black crows indicator.
 *
 * @see [
 * http://www.investopedia.com/terms/t/three_black_crows.asp](http://www.investopedia.com/terms/t/three_black_crows.asp)
 */
/**
 * Constructor.
 *
 * @param series   the bar series
 * @param barCount the number of bars used to calculate the average lower shadow
 * @param factor   the factor used when checking if a candle has a very short
 * lower shadow
 */

class ThreeBlackCrowsIndicator(series: BarSeries?, barCount: Int, factor: Double) : CachedIndicator<Boolean>(series) {
    /**
     * Lower shadow
     */
    private val lowerShadowInd: LowerShadowIndicator

    /**
     * Average lower shadow
     */
    private val averageLowerShadowInd: SMAIndicator

    /**
     * Factor used when checking if a candle has a very short lower shadow
     */
    private val factor: Num
    private var whiteCandleIndex = -1

    init {
        lowerShadowInd = LowerShadowIndicator(series)
        averageLowerShadowInd = SMAIndicator(lowerShadowInd, barCount)
        this.factor = numOf(factor)
    }

    override fun calculate(index: Int): Boolean {
        if (index < 3) {
            // We need 4 candles: 1 white, 3 black
            return false
        }
        whiteCandleIndex = index - 3
        return (barSeries!!.getBar(whiteCandleIndex).isBullish && isBlackCrow(index - 2) && isBlackCrow(index - 1)
                && isBlackCrow(index))
    }

    /**
     * @param index the bar/candle index
     * @return true if the bar/candle has a very short lower shadow, false otherwise
     */
    private fun hasVeryShortLowerShadow(index: Int): Boolean {
        val currentLowerShadow = lowerShadowInd[index]
        // We use the white candle index to remove to bias of the previous crows
        val averageLowerShadow = averageLowerShadowInd.getValue(whiteCandleIndex)
        return currentLowerShadow.isLessThan(averageLowerShadow.times(factor))
    }

    /**
     * @param index the current bar/candle index
     * @return true if the current bar/candle is declining, false otherwise
     */
    private fun isDeclining(index: Int): Boolean {
        val prevBar = barSeries!!.getBar(index - 1)
        val currBar = barSeries.getBar(index)
        val prevOpenPrice = prevBar.openPrice
        val prevClosePrice = prevBar.closePrice
        val currOpenPrice = currBar.openPrice
        val currClosePrice = currBar.closePrice

        // Opens within the body of the previous candle
        return (currOpenPrice!!.isLessThan(prevOpenPrice) && currOpenPrice.isGreaterThan(prevClosePrice) // Closes below the previous close price
                && currClosePrice!!.isLessThan(prevClosePrice))
    }

    /**
     * @param index the current bar/candle index
     * @return true if the current bar/candle is a black crow, false otherwise
     */
    private fun isBlackCrow(index: Int): Boolean {
        val prevBar = barSeries!!.getBar(index - 1)
        val currBar = barSeries.getBar(index)
        return if (currBar.isBearish) {
            if (prevBar.isBullish) {
                // First crow case
                hasVeryShortLowerShadow(index) && currBar.openPrice!!.isLessThan(prevBar.highPrice)
            } else {
                hasVeryShortLowerShadow(index) && isDeclining(index)
            }
        } else false
    }
}